Auth0/SafariAuthenticationCallback.swift

```diff
@@ -22,22 +22,45 @@
 
 import UIKit
 import SafariServices
+#if canImport(AuthenticationServices)
+import AuthenticationServices
+#endif
 
 #if swift(>=3.2)
 @available(iOS 11.0, *)
 class SafariAuthenticationSessionCallback: AuthTransaction {
 
     var state: String?
-    var authSession: SFAuthenticationSession?
     var callback: (Bool) -> Void = { _ in }
 
+    private var authSession: NSObject?
+
     init(url: URL, schemeURL: String, callback: @escaping (Bool) -> Void) {
         self.callback = callback
-        self.authSession = SFAuthenticationSession(url: url, callbackURLScheme: schemeURL) { [unowned self] url, _ in
+        #if canImport(AuthenticationServices)
+        if #available(iOS 12.0, *) {
+            let authSession = ASWebAuthenticationSession(url: url, callbackURLScheme: schemeURL) { [unowned self] url, _ in
+                self.callback(url != nil)
+                TransactionStore.shared.clear()
+            }
+            self.authSession = authSession
+            authSession.start()
+        } else {
+            let authSession = SFAuthenticationSession(url: url, callbackURLScheme: schemeURL) { [unowned self] url, _ in
+                self.callback(url != nil)
+                TransactionStore.shared.clear()
+            }
+            self.authSession = authSession
+            authSession.start()
+        }
+        #else
+        let authSession = SFAuthenticationSession(url: url, callbackURLScheme: schemeURL) { [unowned self] url, _ in
             self.callback(url != nil)
             TransactionStore.shared.clear()
         }
-        self.authSession?.start()
+        self.authSession = authSession
```

Comment:
After re-reading this.. Note that "fallback for 11" and "fallback for 9" have the exact same code. The only thing changing is `SFAuthenticationSession` and `ASWebAuthenticationSession` on 1 of the 3 cases. Please consider re-writing the code so that only that first line is wrapped around the OS check and the remaining is shared.

I imagine something like this is possible:


```swift
self.callback = callback
 #if canImport(AuthenticationServices) && #available(iOS 12.0, *)
      self.authSession = ASWebAuthenticationSession(url: url, callbackURLScheme: schemeURL) { 
      [unowned self] url, _ in
                self.callback(url != nil)
                TransactionStore.shared.clear()
            }
#else  
        self.authSession = SFAuthenticationSession(url: url, callbackURLScheme: schemeURL) { 
       [unowned self] url, _ in
                self.callback(url != nil)
                TransactionStore.shared.clear()
            }
#endif 

self.authSession.start()
```

Reply:
The behaviour is slightly more complex than this, I could assign `self.authSession = ASWeb/SFAuth` However at that point it is now a common foundation type of `NSObject`. This is done so we can reuse one property, it's only real purpose here is to retain the object so the AuthenticationSession doesn't vanish until it's completed.

Problem now is you can't call `self.authSession.start()` as `start()` is not available in `NSObject`.

Reply:
what about then, 

```swift
self.callback = callback
let authSession;
 #if canImport(AuthenticationServices) && #available(iOS 12.0, *)
      authSession = ASWebAuthenticationSession(url: url, callbackURLScheme: schemeURL) { 
      [unowned self] url, _ in
                self.callback(url != nil)
                TransactionStore.shared.clear()
            }
#else  
        authSession = SFAuthenticationSession(url: url, callbackURLScheme: schemeURL) { 
       [unowned self] url, _ in
                self.callback(url != nil)
                TransactionStore.shared.clear()
            }
#endif 
self.authSession = authSession;
authSession.start()
```

Reply:
It's a good idea in theory. However, you're left with the same problem, Swift is strongly type, you must specify the type for your `let authSession` So you are back to the same problem as your previous suggestion.